commit 6890b90f8349a4a7657ed7231d474eb96023c082
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2019-07-01T21:32:14+02:00 (3 months ago)
    
    TRITON-1785 Instance creation should not permit first disk assignment to be lesser than image size

diff --git a/lib/machines.js b/lib/machines.js
index bb0028d4b..ae64e3c1f 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -544,6 +544,12 @@ function getCreateOptions(req) {
                 }
 
 
+                if (disks[0].image_uuid && disks[0].size < img.image_size) {
+                        throw new InvalidArgumentError(
+                            'Size must be greater than or equal to ' +
+                            'image size for the boot disk');
+                }
+
                 if (disksSum > pkg.quota) {
                     throw new InvalidArgumentError(
                         'The size of the disks must not be ' +
diff --git a/test/machines.94.test.js b/test/machines.94.test.js
index c61d89ea0..4cb3a6c50 100644
--- a/test/machines.94.test.js
+++ b/test/machines.94.test.js
@@ -1405,6 +1405,34 @@ test('Package has remaining boot disk', function (suite) {
     suite.end();
 });
 
+
+
+test('Boot disk smaller than image size', function (t) {
+    if (!BHYVE_IMAGE_UUID) {
+        t.ok(true, 'No bhyve images. Test skipped');
+        t.end();
+        return;
+    }
+
+    var obj = {
+        disks: [
+            { size: BHYVE_IMAGE.image_size / 2},
+            { size: 'remaining' }
+        ],
+        image: BHYVE_IMAGE_UUID,
+        name: 'bhyve-remaining-no-space-test-' + process.pid,
+        package: BHYVE_128_FLEXIBLE.uuid
+    };
+
+    CLIENT.post('/my/machines', obj,
+        function createdMachine(err, req, res, body) {
+            t.ok(err);
+            t.equal(err.statusCode, 409);
+            t.equal(body.code, 'InvalidArgument');
+            t.end();
+    });
+});
+
 test('teardown', function (t) {
     common.teardown(CLIENTS, SERVER, function (teardownErr) {
         t.ifError(teardownErr, 'Teardown success');
